src/treetensornetwork/abstracttreetensornetwork.jl

```diff
@@ -0,0 +1,366 @@
+function contract(
+  ψ::AbstractTreeTensorNetwork{V}, root_vertex::V=default_root_vertex(ψ); kwargs...
+) where {V}
+  ψ = copy(ψ)
+  # reverse post order vertices
+  traversal_order = reverse(post_order_dfs_vertices(ψ, root_vertex))
+  return contract(ITensorNetwork(ψ); sequence=traversal_order, kwargs...)
+  # # forward post order edges
+  # ψ = copy(ψ)
+  # for e in post_order_dfs_edges(ψ, root_vertex)
+  #   ψ = contract(ψ, e)
+  # end
+  # return ψ[root_vertex]
+end
+
+function inner(
+  ϕ::AbstractTreeTensorNetwork,
+  ψ::AbstractTreeTensorNetwork;
+  root_vertex=default_root_vertex(ϕ, ψ),
+)
+  ϕᴴ = sim(dag(ψ); sites=[])
+  ψ = sim(ψ; sites=[])
+  ϕψ = ϕᴴ ⊗ ψ
+  # TODO: find the largest tensor and use it as
+  # the `root_vertex`.
+  for e in post_order_dfs_edges(ψ, root_vertex)
+    if has_vertex(ϕψ, (src(e), 2))
+      ϕψ = contract(ϕψ, (src(e), 2) => (src(e), 1))
+    end
+    ϕψ = contract(ϕψ, (src(e), 1) => (dst(e), 1))
+    if has_vertex(ϕψ, (dst(e), 2))
+      ϕψ = contract(ϕψ, (dst(e), 2) => (dst(e), 1))
+    end
+  end
+  return ϕψ[root_vertex, 1][]
+end
+
+function norm(ψ::AbstractTreeTensorNetwork)
+  if isortho(ψ)
+    return norm(ψ[only(ortho_center(ψ))])
+  end
+  return √(abs(real(inner(ψ, ψ))))
+end
+
+# 
+# Utility
+# 
+
+function normalize!(ψ::AbstractTreeTensorNetwork)
+  c = ortho_center(ψ)
+  lognorm_ψ = lognorm(ψ)
+  if lognorm_ψ == -Inf
+    return ψ
+  end
+  z = exp(lognorm_ψ / length(c))
+  for v in c
+    ψ[v] ./= z
+  end
+  return ψ
+end
+
+function normalize(ψ::AbstractTreeTensorNetwork)
+  return normalize!(copy(ψ))
+end
+
+function _apply_to_orthocenter!(f, ψ::AbstractTreeTensorNetwork, x)
+  v = first(ortho_center(ψ))
+  ψ[v] = f(ψ[v], x)
+  return ψ
+end
+
+function _apply_to_orthocenter(f, ψ::AbstractTreeTensorNetwork, x)
+  return _apply_to_orthocenter!(f, copy(ψ), x)
+end
+
+Base.:*(ψ::AbstractTreeTensorNetwork, α::Number) = _apply_to_orthocenter(*, ψ, α)
+
+Base.:*(α::Number, ψ::AbstractTreeTensorNetwork) = ψ * α
+
+Base.:/(ψ::AbstractTreeTensorNetwork, α::Number) = _apply_to_orthocenter(/, ψ, α)
+
+Base.:-(ψ::AbstractTreeTensorNetwork) = -1 * ψ
+
+function LinearAlgebra.rmul!(ψ::AbstractTreeTensorNetwork, α::Number)
+  return _apply_to_orthocenter!(*, ψ, α)
+end
+
+function lognorm(ψ::AbstractTreeTensorNetwork)
+  if isortho(ψ)
+    return log(norm(ψ[only(ortho_center(ψ))]))
+  end
+  lognorm2_ψ = logdot(ψ, ψ)
+  rtol = eps(real(scalartype(ψ))) * 10
+  atol = rtol
+  if !IsApprox.isreal(lognorm2_ψ, Approx(; rtol=rtol, atol=atol))
+    @warn "log(norm²) is $lognorm2_T, which is not real up to a relative tolerance of $rtol and an absolute tolerance of $atol. Taking the real part, which may not be accurate."
+  end
+  return 0.5 * real(lognorm2_ψ)
+end
+
+function loginner(ψ1::TTNT, ψ2::TTNT; kwargs...) where {TTNT<:AbstractTreeTensorNetwork}
+  return logdot(ψ1, ψ2; kwargs...)
+end
+
+# TODO: stick with this traversal or find optimal contraction sequence?
```

Comment:
I think ultimately this should just be designed based on calling a more general `contract` function with a contraction sequence appropriate for a tree network, and we can use a tree traversal as a way to quickly derive a reasonable sequence, and users can pass their own contraction sequence or function for computing a contraction sequence. But that can be left for a future PR.

Reply:
Tracking this issue in #25.